### ISSUE TITLE
GlobalRef.createReaperThreadIfAbsent produces IllegalMonitorStateException

### ISSUE THREAD

**Issue by fra-orolo** [jython/jython#272]
We get this exception when using the Jython logging module in parallel threads in embedded Jython executors.
This causes all embedded Jython interpreters to hang in our system.
```
java.lang.IllegalMonitorStateException: attempt to unlock read lock, not locked by current thread
    at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$Sync.unmatchedUnlockException(ReentrantReadWriteLock.java:448)
    at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$Sync.tryReleaseShared(ReentrantReadWriteLock.java:432)
    at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.releaseShared(AbstractQueuedSynchronizer.java:1094)
    at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.unlock(ReentrantReadWriteLock.java:897)
    at org.python.modules._weakref.GlobalRef.createReaperThreadIfAbsent(GlobalRef.java:343)
    at org.python.modules._weakref.GlobalRef.newInstance(GlobalRef.java:235)
    at org.python.modules._weakref.ReferenceType.weakref___new__(ReferenceType.java:36)
    at org.python.modules._weakref.ReferenceType$exposed___new__.new_impl(Unknown Source)
    at org.python.core.PyType.invokeNew(PyType.java:1119)
```

The problem is in the implementation of this method in GlobalRef:
https://github.com/jython/jython/blob/89a1733554b9351d2ce0d409f711e036cea776e0/src/org/python/modules/_weakref/GlobalRef.java#L327-L346

The readLock is aquired in L328 and intermediately released in L331. 
Normally it is re-aquired in the first finally block L337, but only if the repeated if() in L333 finds still no active reaperThread.
Otherwise the read-lock is not re-aquired, but still bluntly unlocked in the last finally L343.

The code would be more correct, if the method would be rewritten like this:
```
    private static void createReaperThreadIfAbsent() {
        reaperLock.readLock().lock();
        try {
            if (reaperThread == null || !reaperThread.isAlive()) {
                reaperLock.readLock().unlock();
                reaperLock.writeLock().lock();
                try {
                    if (reaperThread == null || !reaperThread.isAlive()) {
                        initReaperThread();
                    }
                } finally {
                    reaperLock.readLock().lock();
                    reaperLock.writeLock().unlock();
                }
            }
        } finally {
            reaperLock.readLock().unlock();
        }
    }

```

**Comment by Stewori** [comment]
Thanks for this work. It looks good to me and given that it passes tests I would accept the mentioned PR subject to the following request (unless @jeff5 vetoes):
Although maybe unrelated to your described issue, on this occasion could you please make ``reaperThread`` volatile? See https://www.cs.umd.edu/%7Epugh/java/memoryModel/DoubleCheckedLocking.html.
Did you sign the contributor agreement?
Then, if not yet done please feel free to add yourself to the contributors list (as part of the PR).

**Comment by fra-orolo** [comment]
Hi,
I added the volatile reaperThread.
I have not signed the contributor agreement yet. I found a text on python.org, is there a (electronic) process for this?
Where should I send it?

**Comment by Stewori** [comment]
Thank you! I think the CLA should be signed with the PSF under https://www.python.org/psf/contrib/contrib-form/ as the PSF governs Jython. It has been some time since I last dealt with this. A quick search lead me to this discussion #74 which confirms the contrib-form link.

**Comment by fra-orolo** [comment]
Ok, thanks for the link. I signed to be a contributor.

**Comment by Stewori** [comment]
Do you want to add yourself to the list of people who contributed to Jython in the ACKNOWLEDGMENTS file? (You don't have to.) If you want to be listed, please add yourself as part of your PR.

**Comment by fra-orolo** [comment]
Hi Stefan,
thanks again for the offer, but no: I'm fine with being git-blame'd for the few lines that I shuffled.
Best regards
Martin

**Comment by Stewori** [comment]
Fixed as of #273. Martin, thanks again for this fix!
